### ISSUE TITLE
VQ-VAE Loss Function 

### ISSUE THREAD

**Issue by Arstanley** [rosinality/vq-vae-2-pytorch#3]
I noticed that in train_vqvae.py, there are only two terms for the loss function; however, there appears to be three terms in the paper. I suppose the second term in the code refers to the commitment loss, but if that is the case, how do we update the coding book then? The code works very well, but I am just wondering about this finding.

> out, latent_loss = model(img)
        recon_loss = criterion(out, img)
        loss = recon_loss + latent_loss_weight * latent_loss.mean()
        loss.backward()

**Comment by pclucas14** [comment]
The codebook is updated using Exponential Moving Averages. The update equations are in the appendix of the original paper (and implemented [here](https://github.com/rosinality/vq-vae-2-pytorch/blob/master/vqvae.py#L51))

**Comment by Arstanley** [comment]
Thank you so much!
